The main goal of association mining is finding underlying relationship between items. The most common question is asked in association mining- which are the items purchased together? To answer this question two popular methods are used listed below. This analysis is useful to find likelyhood of purchasing item Y if another item X is bought and measure based on confidence

Benefits of Association mining:
1. if item A and B are bought together more frequently then several steps can be taken to increase the profit. 
2. A and B can be placed together so that when a customer buys one of the product he doesn't have to go far away to buy the other product.
3. People who buy one of the products can be targeted through an advertisement campaign to buy the other.
4. Collective discounts can be offered on these products if the customer buys both of them.
5. Both A and B can be packaged together.

Some terminoloy in association mining.

confidencd (Y/X)=(total transaction of X and Y together)/(total transaction of X)

Support(Y) = (Transactions containing Y)/(Total Transactions)

Lift: likelyhood of buying X and Y together is N times than jusy buying X. Lift <=1 no association between items. 

Lift(X→Y) = (Confidence (X→Y))/(Support (Y))


1. Apriori: very simple algorithm but has huge overhead because database needs to be scanned again and again.
2. FP-Growth

Lets see an example with implementation
=====

In [35]:
from apyori import apriori
import pandas as pd

In [42]:
data=pd.read_csv("dataset/store_data.csv", header=None)
print(data.shape[0])

7501


In [43]:
min_sale_per_week=5
min_length=3 # number of items in the rule
min_support=min_sale_per_week*7/data.shape[0]
transaction = []
for i in range(0, data.shape[0]):
    transaction.append([str(data.values[i,j]) for j in range(0, 20)])

app_ass=apriori(transaction, min_support=min_support, min_confidence=0.2, min_lift=4, min_length=min_length, use_colnames=True)

app_ass_result=list(app_ass)
print(app_ass_result[0])

RelationRecord(items=frozenset({'escalope', 'pasta'}), support=0.005865884548726837, ordered_statistics=[OrderedStatistic(items_base=frozenset({'pasta'}), items_add=frozenset({'escalope'}), confidence=0.3728813559322034, lift=4.700811850163794)])


In [44]:
for item in app_ass_result:

    # first index of the inner list
    # Contains base item and add item
    pair = item[0] 
    items = [x for x in pair]
    print("Rule: " + items[0] + " -> " + items[1])

    #second index of the inner list
    print("Support: " + str(item[1]))

    #third index of the list located at 0th
    #of the third index of the inner list

    print("Confidence: " + str(item[2][0][2]))
    print("Lift: " + str(item[2][0][3]))
    print("=====================================")

Rule: escalope -> pasta
Support: 0.005865884548726837
Confidence: 0.3728813559322034
Lift: 4.700811850163794
Rule: olive oil -> whole wheat pasta
Support: 0.007998933475536596
Confidence: 0.2714932126696833
Lift: 4.122410097642296
Rule: shrimp -> pasta
Support: 0.005065991201173177
Confidence: 0.3220338983050847
Lift: 4.506672147735896
Rule: nan -> escalope
Support: 0.005865884548726837
Confidence: 0.3728813559322034
Lift: 4.700811850163794
Rule: ground beef -> spaghetti
Support: 0.006399146780429276
Confidence: 0.3934426229508197
Lift: 4.004359721511667
Rule: olive oil -> nan
Support: 0.007998933475536596
Confidence: 0.2714932126696833
Lift: 4.13077198425009
Rule: nan -> shrimp
Support: 0.005065991201173177
Confidence: 0.3220338983050847
Lift: 4.515095833993347
Rule: ground beef -> spaghetti
Support: 0.006399146780429276
Confidence: 0.3934426229508197
Lift: 4.004359721511667


Rule: escalope -> pasta

Support: 0.005865884548726837=total transaction contains (escalope)/total transaction

Confidence: 0.3728813559322034=37% of the transaction contained escalope and pasta together among all transaction for escalope 

Lift: 4.700811850163794= likelyhood of escalope and pasta will be sold together 4 times more than escalope alone 